lightning/src/ln/data_persister.rs

```diff
@@ -0,0 +1,221 @@
+/// revocation can be persisted, then unintentionally broadcasting a revoked
+/// transaction and losing money. This is a risk because previous channel states
+/// are toxic, so it's important that whatever channel state is persisted is
+/// kept up-to-date.
+///
+/// Upon calls to update_channel_data, implementers of this trait should either
+/// persist the ChannelMonitorUpdate or the ChannelMonitor. If an implementer
+/// chooses to persist the updates only, they need to make sure that all the
+/// updates are applied to the ChannelMonitors *before* the set of channel
+/// monitors is given to the ChainMonitor at startup time (e.g., all monitors
+/// returned by `load_channel_data` must be up-to-date). If full ChannelMonitors
+/// are persisted, then there is no need to persist individual updates.
+///
+/// Note that there could be a performance tradeoff between persisting complete
+/// channel monitors on every update vs. persisting only updates and applying
+/// them in batches.
+///
+/// Given multiple backups, situations may arise where one or more backup sources
+/// have fallen behind or disagree on the current state. Because these backup
+/// sources don't have any transaction signing/broadcasting duties and are only
+/// supposed to be persisting bytes of data, backup sources may be considered
+/// "in consensus" if a majority of them agree on the current state and are on
+/// the highest known commitment number. See individual methods for more info.
+pub trait ChannelDataPersister: Send + Sync {
+	/// The concrete type which signs for transactions and provides access to our channel public
+	/// keys.
+	type Keys: ChannelKeys;
+
+	/// Persist a new channel's data. The majority of backups should agree on a
+	/// channel's state and be on the latest [`update_id`]. The data
+	/// can be stored with any file name/path, but the identifier provided is the
+	/// channel's outpoint.
+	///
+	/// See [`ChannelMonitor::write_for_disk`] for writing out a ChannelMonitor.
+	///
+	/// [`update_id`]: struct.ChannelMonitorUpdate.html#structfield.update_id
+	/// [`ChannelMonitor::write_for_disk`]: ../../chain/channelmonitor/struct.ChannelMonitor.html#method.write_for_disk
+	fn persist_channel_data(&self, id: OutPoint, data: &ChannelMonitor<Self::Keys>) -> Result<(), ChannelMonitorUpdateErr>;
+
+	/// Update one channel's data. The provided ChannelMonitor has already
+	/// applied the given update.
+	///
+	/// Note that on every update, you **must** persist either the
+	/// ChannelMonitorUpdate or the updated monitor itself to disk/backups.
+	/// Otherwise, there is risk of situations such as revoking a transaction,
+	/// then crashing before this revocation can be persisted, then
+	/// unintentionally broadcasting a revoked transaction and losing money. This
+	/// is a risk because previous channel states are toxic, so it's important
+	/// that whatever channel state is persisted is kept up-to-date.
```

Comment:
This looks to repeat what's in the struct documentation. To avoid these going out-of-sync, only state this in one place, possibly summarizing in and/or referencing this from the other place.

Reply:
thanks, I put the fixes for these two comments in the `documentation fixes` commit.